kimchi/src/snarky/checked_runner.rs

```diff
@@ -0,0 +1,424 @@
+//! The circuit-generation and witness-generation logic.
+
+use ark_ff::PrimeField;
+
+use crate::{
+    circuits::gate::CircuitGate,
+    curve::KimchiCurve,
+    loc,
+    snarky::{
+        boolean::Boolean,
+        constraint_system::{BasicSnarkyConstraint, KimchiConstraint, SnarkyConstraintSystem},
+        cvar::CVar,
+        traits::SnarkyType,
+    },
+};
+
+use super::constants::Constants;
+
+/// A wrapper around [BasicSnarkyConstraint] and [KimchiConstraintSystem] that allows for an optional label (for debugging).
+pub struct AnnotatedConstraint<F: PrimeField> {
+    annotation: Option<&'static str>,
+    constraint: Constraint<F>,
+}
+
+impl<F> AnnotatedConstraint<F>
+where
+    F: PrimeField,
+{
+    /// In witness generation, this checks if the constraint is satisfied by some witness values.
+    pub fn check_constraint(&self, env: &impl WitnessGeneration<F>) {
+        match &self.constraint {
+            Constraint::BasicSnarkyConstraint(c) => c.check_constraint(env),
+            Constraint::KimchiConstraint(c) => c.check_constraint(env),
+        }
+    }
+}
+
+/// An enum that wraps either a [BasicSnarkyConstraint] or a [KimchiConstraintSystem].
+// TODO: we should get rid of this once basic constraint system is gone
+pub enum Constraint<F: PrimeField> {
+    /// Old R1CS-like constraints.
+    BasicSnarkyConstraint(BasicSnarkyConstraint<CVar<F>>),
+
+    /// Custom gates in kimchi.
+    KimchiConstraint(KimchiConstraint<CVar<F>, F>),
+}
+
+/// The two ways to create a [SnarkyType].
+#[derive(Default)]
+pub enum TypeCreation {
+    /// This will create constraints in the constraint system to ensure that the type is valid.
+    /// This is done by calling [SnarkyType::check] on a snarky type.
+    #[default]
+    Checked,
+
+    /// This will create a new type without any constraints.
+    /// This should only be used internally when we know the type is already constrained to be valid.
+    Unsafe,
+}
+
+/// The mode in which the [RunState] is running.
+#[derive(Default, Clone, Copy)]
+pub enum Mode {
+    /// This will construct the kimchi gates for the circuit.
+    #[default]
+    CircuitGeneration,
+
+    /// This will construct the execution trace (or witness) for the circuit.
+    WitnessGeneration,
+}
+
+/// The state used when compiling a circuit in snarky, or used in witness generation as well.
+#[derive(Default)]
+pub struct RunState<F>
+where
+    F: PrimeField,
+{
+    /// The constraint system used to build the circuit. Why is this an option?
+    system: Option<SnarkyConstraintSystem<F>>,
+
+    /// The public input of the circuit used in witness generation.
+    public_input: Vec<F>,
+
+    /// The private input of the circuit used in witness generation. Still not sure what that is, or why we care about this.
+    private_input: Vec<F>,
+
+    /// If this is set to false, constraints won't be added to the system. Why???
```

Comment:
Sometimes you want to run a circuit in a lightweight 'run this snarky code and just give me the result' form, so that you can reuse it as if it's native code.
This is most useful for testing, where you want to test that the in-circuit implementation is consistent with some native one.

Reply:
thanks!

I think we should have an enum that lists all of the mode in which we can run

Reply:
btw this is actually to test the witness :o